ports/unix/unix_mphal.c

```diff
@@ -214,13 +217,17 @@ mp_uint_t mp_hal_ticks_us(void) {
     return tv.tv_sec * 1000000 + tv.tv_usec;
     #endif
 }
+#endif
 
+#ifndef mp_hal_time_ns
 uint64_t mp_hal_time_ns(void) {
     struct timeval tv;
     gettimeofday(&tv, NULL);
     return (uint64_t)tv.tv_sec * 1000000000ULL + (uint64_t)tv.tv_usec * 1000ULL;
 }
+#endif
 
+#ifndef mp_hal_delay_ms
```

Comment:
Since this one is implemented in terms of mp_hal_ticks_ms, souldn't it be a bit tedioud to have to reimplement this?

Related (since this uses usleep): shouldn't the sleep function(s) also be made configurable? If the (virtual) timers aren't in sync with the sleep calls that's not unlikely to cause problems.

Reply:
For a virtual clock, I would `#define` all of the functions to replaced them with the virtual implementation so `usleep()` would never be called.

Reply:
Unless usleep is used elsewhere in the code without a hal wrapper?

Reply:
> For a virtual clock, I would #define all of the functions to replaced them with the virtual implementation so usleep() would never be called

Makes sense, just a bit of work to have to reimplement mp_hal_delay_ms with the pll hook etc.

> Unless usleep is used elsewhere in the code without a hal wrapper?

Sort of, mod_time_sleep uses `select` with a timeout.


I assume you're doing this towards fixing e.g. #8301 (haven't had time anymore for that one), then it would make sense to do it immediately in windows_mphal.c as well I think, probably best on top of https://github.com/stinos/micropython/commit/aa4688fde8fd2a207a0d3e6bfee82c299875722b

Reply:
> Makes sense, just a bit of work to have to reimplement mp_hal_delay_ms with the pll hook etc.

if `MICROPY_EVENT_POLL_HOOK` is defined then `usleep()` isn't called (provided a new definition of `MICROPY_EVENT_POLL_HOOK` is given that doesn't call `usleep()`. 

But maybe it would be better if we replace the existing calls to `usleep()` with `mp_hal_delay_us()`?

> Unless usleep is used elsewhere in the code without a hal wrapper?

It looks like there are also `usleep()` calls in `ports/unix/mpbtstackport*`.

> Sort of, mod_time_sleep uses `select` with a timeout.

Good point. We are mostly just testing external code that uses `mp_hal_delay_ms()` so far so I missed this.

> I assume you're doing this towards fixing e.g. #8301

My primary motivation is for some out of tree code, but yes, it could be helpful for that too.

Reply:
>> Sort of, mod_time_sleep uses select with a timeout.

> Good point. We are mostly just testing external code that uses mp_hal_delay_ms() so far so I missed this.

I suppose we could make a new, optional `modtime.c` that just calls the `mp_hal` functions instead.

Reply:
> But maybe it would be better if we replace the existing calls to usleep() with mp_hal_delay_us()?

That would probably be a good starting place yes.

Didn't really look into it so not super clear why mod_time_sleep even exists, but I think that `select` in modtime could either be replaced with `mp_hal_delay_us`, or implemented in terms of moduselect, or maybe even just removed completely and be replaced by `mp_hal_delay_ms` because after all that one essentially does the same, no? (apart from the MP_THREAD_GIL_EXIT/ENTER calls, which seems important i.e. makes me wonder whether the `mp_hal_delay_ms` implementation shouldn't also have those)

Reply:
I started https://github.com/micropython/micropython/issues/8977 to continue this discussion. There are some good points here, but I think it is independent of this pull request.